nifi-external/nifi-kafka-connect/nifi-kafka-connector/src/main/java/org/apache/nifi/kafka/connect/StatelessKafkaConnectorWorkingDirectoryUtil.java

```diff
@@ -0,0 +1,98 @@
+
+public class StatelessKafkaConnectorWorkingDirectoryUtil {
+
+    public static final String NAR_UNPACKED_SUFFIX = "nar-unpacked";
+    public static final String HASH_FILENAME = "nar-digest";
+    private static final Logger logger = LoggerFactory.getLogger(StatelessKafkaConnectorWorkingDirectoryUtil.class);
+
+    /**
+     * Goes through the nar/extensions and extensions directories within the working directory
+     * and deletes every directory whose name ends in "nar-unpacked" and does not have a
+     * "nar-digest" file in it.
+     * @param workingDirectory File object pointing to the working directory.
+     */
+    public static void checkWorkingDirectoryIntegrity(final File workingDirectory) {
+        purgeIncompleteUnpackedNars(new File(new File(workingDirectory, "nar"), "extensions"));
+        purgeIncompleteUnpackedNars(new File(workingDirectory, "extensions"));
+    }
+
+    /**
+     * Receives a directory as parameter and goes through every directory within it that ends in
+     * "nar-unpacked". If a directory ending in "nar-unpacked" does not have a file named
+     * "nar-digest" within it, it gets deleted with all of its contents.
+     * @param directory A File object pointing to the directory that is supposed to contain
+     *                  further directories whose name ends in "nar-unpacked".
+     */
+    public static void purgeIncompleteUnpackedNars(final File directory) {
+        final File[] unpackedDirs = directory.listFiles(file -> file.isDirectory() && file.getName().endsWith(NAR_UNPACKED_SUFFIX));
+        if (unpackedDirs == null || unpackedDirs.length == 0) {
+            logger.debug("Found no unpacked NARs in {}", directory);
+            logger.debug("Directory contains: {}", Arrays.deepToString(directory.listFiles()));
+            return;
+        }
+
+        for (final File unpackedDir : unpackedDirs) {
+            final File narHashFile = new File(unpackedDir, HASH_FILENAME);
+            if (!narHashFile.exists()) {
+                purgeDirectory(unpackedDir);
+            } else {
+                logger.debug("Already successfully unpacked {}", unpackedDir);
+            }
+        }
+    }
+
+    /**
+     * Delete a directory with all of its contents.
+     * @param directory The directory to be deleted.
+     */
+    public static void purgeDirectory(final File directory) {
+        if (directory.exists()) {
+            deleteRecursively(directory);
+            logger.debug("Cleaned up {}", directory);
+        }
+    }
+
+    private static void deleteRecursively(final File fileOrDirectory) {
+        if (fileOrDirectory.isDirectory()) {
+            final File[] files = fileOrDirectory.listFiles();
+            if (files != null) {
+                for (final File file : files) {
+                    deleteRecursively(file);
+                }
+            }
+        }
+        deleteOrDebug(fileOrDirectory);
+    }
+
+    private static void deleteOrDebug(final File file) {
```

Comment:
Recommend renaming to `deleteQuietly`.
```suggestion
    private static void deleteQuietly(final File file) {
```